@tgamblin here's what I've been trying. 

The spec is getting chopped up in the parser though. Not sure I understand the significance of calling `push_tokens` at line 4101 of `spec.py` but it seems like the spec is getting chopped up in an undesirable manner somewhere there, or when it goes back to `do_parse`.

```
        parsed = (" ".join(str(spec) for spec in output))
>       assert expected == parsed
E       AssertionError: assert 'yaml-cpp@0.1.8%gcc' == '8%gcc'
E         - yaml-cpp@0.1.8%gcc
E         + 8%gcc
```

--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -3945,7 +3945,8 @@ class SpecLexer(spack.parse.Lexer):
 
             # Filenames match before identifiers, so no initial filename
             # component is parsed as a spec (e.g., in subdir/spec.yaml)
-            (r'[/\w.-]+\.yaml$', lambda scanner, v: self.token(FILE, v)),
+            (r'[/\w.-]*/[/\w/-]+\.yaml$',
+             lambda scanner, v: self.token(FILE, v)),
 
             # Hash match after filename. No valid filename can be a hash
             # (files end w/.yaml), but a hash can match a filename prefix.
@@ -4096,23 +4097,6 @@ class SpecParser(spack.parse.Parser):
         """
         path = self.token.value
 
-        # don't treat builtin.yaml, builtin.yaml-cpp, etc. as filenames
-        if re.match(spec_id_re, path):
-            self.push_tokens([spack.parse.Token(ID, self.token.value)])
-            return None
-
-        # Special case where someone omits a space after a filename. Consider:
-        #
-        #     libdwarf^/some/path/to/libelf.yamllibdwarf ^../../libelf.yaml
-        #
-        # The error is clearly an omitted space. To handle this, the FILE
-        # regex admits text *beyond* .yaml, and we raise a nice error for
-        # file names that don't end in .yaml.
-        if not path.endswith(".yaml"):
-            raise SpecFilenameError(
-                "Spec filename must end in .yaml: '{0}'".format(path))
-
-        # if we get here, we're *finally* interpreting path as a filename
         if not os.path.exists(path):
             raise NoSuchSpecFileError("No such spec file: '{0}'".format(path))
 
@@ -4458,11 +4442,7 @@ class NoSuchHashError(spack.error.SpecError):
             % hash)
 
 
-class SpecFilenameError(spack.error.SpecError):
-    """Raised when a spec file name is invalid."""
-
-
-class NoSuchSpecFileError(SpecFilenameError):
+class NoSuchSpecFileError(spack.error.SpecError):
     """Raised when a spec file doesn't exist."""
 
 

--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -3945,7 +3945,7 @@ class SpecLexer(spack.parse.Lexer):
 
             # Filenames match before identifiers, so no initial filename
             # component is parsed as a spec (e.g., in subdir/spec.yaml)
-            (r'[/\w.-]+\.yaml[^\b]*', lambda scanner, v: self.token(FILE, v)),
+            (r'[/\w.-]+\.yaml$', lambda scanner, v: self.token(FILE, v)),
 
             # Hash match after filename. No valid filename can be a hash
             # (files end w/.yaml), but a hash can match a filename prefix.
@@ -4094,10 +4094,10 @@ class SpecParser(spack.parse.Parser):
            we backtrack from spec_from_file() and treat them as spec names.
 
         """
-        path = self.token.value.split(" ")[0]
+        path = self.token.value
 
         # don't treat builtin.yaml, builtin.yaml-cpp, etc. as filenames
-        if re.match(r'\w[\w.%@-]*$', path):
+        if re.match(spec_id_re, path):
             self.push_tokens([spack.parse.Token(ID, self.token.value)])
             return None
 

--- a/lib/spack/spack/test/spec_syntax.py
+++ b/lib/spack/spack/test/spec_syntax.py
@@ -19,7 +19,7 @@ from spack.spec import SpecParseError, RedundantSpecError
 from spack.spec import AmbiguousHashError, InvalidHashError, NoSuchHashError
 from spack.spec import DuplicateArchitectureError
 from spack.spec import DuplicateDependencyError, DuplicateCompilerSpecError
-from spack.spec import SpecFilenameError, NoSuchSpecFileError
+from spack.spec import NoSuchSpecFileError
 from spack.variant import DuplicateVariantError
 
 
@@ -192,27 +192,27 @@ class TestSpecSyntax(object):
         self.check_parse(
             "yaml-cpp@0.1.8%intel@12.1"
             " ^boost@3.1.4")
-        tempSpec = r"builtin.yaml-cpp%gcc"
+        tempspec = r"builtin.yaml-cpp%gcc"
         self.check_parse(
-            tempSpec.strip("builtin."),
-            spec=tempSpec)
-        tempSpec = r"testrepo.yaml-cpp%gcc"
+            tempspec.strip("builtin."),
+            spec=tempspec)
+        tempspec = r"testrepo.yaml-cpp%gcc"
         self.check_parse(
-            tempSpec.strip("testrepo."),
-            spec=tempSpec)
-        tempSpec = r"builtin.yaml-cpp@0.1.8%gcc"
+            tempspec.strip("testrepo."),
+            spec=tempspec)
+        tempspec = r"builtin.yaml-cpp@0.1.8%gcc"
         self.check_parse(
-            tempSpec.strip("builtin."),
-            spec=tempSpec)
-        tempSpec = r"builtin.yaml-cpp@0.1.8%gcc@7.2.0"
+            tempspec.strip("builtin."),
+            spec=tempspec)
+        tempspec = r"builtin.yaml-cpp@0.1.8%gcc@7.2.0"
         self.check_parse(
-            tempSpec.strip("builtin."),
-            spec=tempSpec)
-        tempSpec = r"builtin.yaml-cpp@0.1.8%gcc@7.2.0" \
+            tempspec.strip("builtin."),
+            spec=tempspec)
+        tempspec = r"builtin.yaml-cpp@0.1.8%gcc@7.2.0" \
             r" ^boost@3.1.4"
         self.check_parse(
-            tempSpec.strip("builtin."),
-            spec=tempSpec)
+            tempspec.strip("builtin."),
+            spec=tempspec)
 
     def test_canonicalize(self):
         self.check_parse(

--- a/lib/spack/spack/test/spec_syntax.py
+++ b/lib/spack/spack/test/spec_syntax.py
@@ -649,13 +649,6 @@ class TestSpecSyntax(object):
             'mvapich_foo ./libelf.yaml',
         ])
 
-    def test_nice_error_for_no_space_after_spec_filename(self):
-        """Ensure that omitted spaces don't give weird errors about hashes."""
-        self._check_raises(SpecFilenameError, [
-            '/bogus/path/libdwarf.yamlfoobar',
-            'libdwarf^/bogus/path/libelf.yamlfoobar ^/path/to/bogus.yaml',
-        ])
-
     @pytest.mark.usefixtures('config')
     def test_yaml_spec_not_filename(self, mock_packages, tmpdir):
         with pytest.raises(spack.repo.UnknownPackageError):

